### ISSUE TITLE
Question

### ISSUE THREAD

**Issue by zero77** [gorhill/uMatrix#825]
To keep the “my rules” page tidy is it possible to have more than one request type per rule.
google.com google.com xhr,frame,other allow
thanks

**Comment by gorhill** [comment]
Please provide sensible title to issue you open, these are used for others to lookup issues, and it helps by not burdening me with the task of having to edit your issue. Is the following excerpt not clear in [CONTRIBUTING](https://github.com/gorhill/uMatrix/blob/master/CONTRIBUTING.md)?

> you benefit from the carefully crafted code, return the favor by carefully crafting issues/bug reports.

**Comment by zero77** [comment]
Apologies, I will take this into consideration in the future.

**Comment by elypter** [comment]
this makes things a little more complicated syntax wise because then multiple ways of writing rules exist but it makes the list a lot smaller. what i think that would be even more effient if you go that route would be meta rules like "all-except-cookies" or "video-frame" which is basically a generic row that can be applied to any domain you like and that can be changed globally. if umatrix one day become more fine grained this could prove handy. they would be defined with a directive and can then be used like regular request types.

**Comment by gorhill** [comment]
Rules have to map to `src des type action` internally, this won't change, this is essentially uMatrix at its core. The suggestion here translate very well back and forth to uMatrix's internal syntax. Anything else that does not translate well into `src des type action` won't work and will just make the code unduly complicated -- not a good thing for stability.

**Comment by elypter** [comment]
this would still have the same syntax. just more keywords for type. for example:

`yahoo.com search.yahoo.com except-cookies allow`

this would allow every request on yahoo.com to search.yahoo.com except cookies

for this to work you would have to add a new directive:
`custom-type: type-name type action`

for example:
`custom-type: except-cookie * allow`
`custom-type: except-cookie cookie block`

this could even be expanded for a whole scope by adding destinations to the syntax. this way you could reuse a ruleset made for one website for another or make templates for general types of websites and only adjust whats different between them. for example to allow embedded youtube videos easily on a set of sites without having to allow google globally.

`template: type-name des type action`

for example:
`template: except-cookie 1st-party * allow`
`template: except-cookie * cookie block`
`template: except-cookie 1st-party cookie block`

i dont know if it is really needed and its a big change but i think if that if the syntax should be extended this would be a good and flexible way.

**Comment by zero77** [comment]
I was thinking of something along the lines of:
google.com mail.google.com cookie,xhr,other allow

**Comment by ArchangeGabriel** [comment]
Yeah, the idea is that the `type` field could be a list of comma separated types instead of only one.

**Comment by zero77** [comment]
@ArchangeGabriel 
yes, exactly.

**Comment by gorhill** [comment]
> I was thinking of something along the lines of:

That was my understanding of your issue, and what I was planning to do, including keeping the types ordered as per current order in the matrix UI.

**Comment by zero77** [comment]
@gorhill 
thanks.

**Comment by mikhaelkh** [comment]
As I see it, the main goal of this request is to make "My rules" tab easier to work with.
But grouping is not the best way to make it happen. Also, it complicates the implementation of my ideas.
I think the best way of improving usability is:
1) Search. Hide rules that doesn't contain given string.
2) Hide the same rules in Permanent and Temporary halves. Useful to review what you've changed before closing the browser.

**Comment by zero77** [comment]
That’s a good idea but it’s not really what I was after in this request.
I like the my rules tab and I find it easy to work with.

But, I am finding its getting a bit messy as I have lots of rules and in particular multiple rules for the same domains. 

So I think combining rules for the same domain would solve this and make the my rules tab a lot neater and easer to work with as a result.
